File: apps/desktop/src/ipc/playground.ts
/**
 * 游乐场：会话、消息、图片，以及发一轮 / 出一批图 / 停。
 *
 * 与 `commands/playground.rs` 一一对应；类型是 `nexus_playground::model` 的镜像。
 * 前端每次只交「thread_id + 新的一句话」，历史在 Rust 侧从库里拼；流式回字走
 * `playground://chat` 事件，帧形状与「试一下」同一套（`TryFrame`）。
 * 图片字节不过 IPC：凭 `ImageRef.id` 拼成 `nexus-image://` 地址交给 `<img>`。
 */
import { convertFileSrc, invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";
import type { TryFrame, TryUsage } from "./models";

export type Kind = "chat" | "image" | "video";
export type Role = "user" | "assistant";

export interface Thread {
  id: string;
  kind: Kind;
  /** 空串 = 还没标题（第一条消息进来时用它的开头）。 */
  title: string;
  model: string;
  createdAt: string;
  updatedAt: string;
}

export interface ThreadSummary extends Thread {
  messageCount: number;
  /** 最后一条有内容的消息的开头。 */
  preview: string | null;
  /** 图片会话的封面：最近一张图的 id。 */
  coverImageId: string | null;
}

export interface ImageRef {
  id: string;
  messageId: string;
  mime: string;
  width: number | null;
  height: number | null;
  bytes: number;
  /** 请求时要的规格（`1024x1024`）。 */
  size: string | null;
  createdAt: string;
}

export interface Message {
  id: string;
  threadId: string;
  seq: number;
  role: Role;
  content: string;
  thinking: string | null;
  /** 产出这条回复时请求的模型（只有 assistant 有）。 */
  model: string | null;
  /** 上游实际路由到的模型。 */
  routed: string | null;
  usage: TryUsage | null;
  /** 这一轮失败的原因；有内容也可能有它（中途断掉、被停止）。 */
  error: string | null;
  durationMs: number | null;
  ttftMs: number | null;
  createdAt: string;
  images: ImageRef[];
}

export interface ThreadDetail {
  thread: Thread;
  messages: Message[];
}

/** 资产页里的一张图：图片本身 + 它出自哪个会话、哪个模型、哪句提示词。 */
export interface Asset extends ImageRef {
  threadId: string;
  threadTitle: string;
  model: string | null;
  prompt: string | null;
}

/** 这个会话正在跑的那一次请求，连同攒到现在的半截回复。 */
export interface ActiveRun {
  requestId: string;
  threadId: string;
  kind: Kind;
  text: string;
  thinking: string;
  routed: string | null;
  startedMsAgo: number;
}

export interface ImageRequest {
  prompt: string;
  size: string | null;
  n: number;
}

/** 一次生视频。首帧图的字节随命令上去（不带 `data:` 前缀）。 */
export interface VideoRequest {
  prompt: string;
  imageBase64: string | null;
  imageMime: string | null;
  /** 1–15 秒。 */
  duration: number | null;
  /** `16:9` 这类。 */
  aspectRatio: string | null;
  /** `480p` / `720p` / `1080p`。 */
  resolution: string | null;
}

/** 视频成片也存在 `ImageRef` 里：靠 MIME 认。 */
export function isVideoRef(ref: Pick<ImageRef, "mime">): boolean {
  return ref.mime.startsWith("video/");
}

/**
 * 随一句话发上去的图。字节在这里**要**过 IPC（只有这一趟：Rust 落盘之后前端就凭 id 取图）。
 * `dataBase64` 不带 `data:` 前缀。
 */
export interface ChatAttachment {
  name: string;
  mime: string;
  dataBase64: string;
}

export const PLAYGROUND_CHAT_EVENT = "playground://chat";

export const playground = {
  threads: (kind?: Kind) => invoke<ThreadSummary[]>("playground_threads", { kind: kind ?? null }),
  thread: (id: string) => invoke<ThreadDetail>("playground_thread", { id }),
  createThread: (kind: Kind, model: string) => invoke<Thread>("playground_thread_create", { kind, model }),
  renameThread: (id: string, title: string) => invoke<Thread>("playground_thread_rename", { id, title }),
  /** 换这个会话下一次发送用的模型。 */
  setTarget: (id: string, model: string) => invoke<Thread>("playground_thread_set_target", { id, model }),
  deleteThread: (id: string) => invoke<void>("playground_thread_delete", { id }),
  deleteMessage: (id: string) => invoke<void>("playground_message_delete", { id }),
  active: (threadId: string) => invoke<ActiveRun | null>("playground_active", { threadId }),
  /**
   * 发一轮。`prompt` 缺省 = 重新生成末尾那条回复（那时不能带附件）。Promise 在回复
   * **落库后**才 resolve，中途的字在 `playground://chat` 事件里，按 `requestId` 认。
   */
  send: (requestId: string, threadId: string, prompt?: string, attachments?: ChatAttachment[]) =>
    invoke<Message>("playground_chat_send", { requestId, threadId, prompt: prompt ?? null, attachments: attachments ?? [] }),
  stop: (requestId: string) => invoke<boolean>("playground_stop", { requestId }),
  generateImage: (requestId: string, threadId: string, request: ImageRequest) =>
    invoke<Message>("playground_image_generate", { requestId, threadId, request }),
  /** 出一段视频。Promise 等上游做完、成片落盘后才 resolve（可能几分钟）；中途 `stop` 就不再等。 */
  generateVideo: (requestId: string, threadId: string, request: VideoRequest) =>
    invoke<Message>("playground_video_generate", { requestId, threadId, request }),
  revealImage: (id: string) => invoke<void>("playground_image_reveal", { id }),
  /** 全部生成过的图，新的在前。 */
  assets: () => invoke<Asset[]>("playground_assets"),
  /** 删一张图（连文件）；消息留着。 */
  deleteImage: (id: string) => invoke<void>("playground_image_delete", { id }),
  listen: (cb: (f: TryFrame) => void): Promise<UnlistenFn> => listen<TryFrame>(PLAYGROUND_CHAT_EVENT, (e) => cb(e.payload)),
};

/** 一张图在 WebView 里的地址。macOS / Linux 是 `nexus-image://localhost/<id>`，Windows 是 `http://nexus-image.localhost/<id>`。 */
export function imageSrc(id: string): string {
  return convertFileSrc(id, "nexus-image");
}

/** 一次发送的请求 id：时间戳 + 随机尾巴，同一页里连点两次也分得清。 */
export function newRequestId(): string {
  return `${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 8)}`;
}

File: apps/desktop/src-tauri/src/commands/playground.rs
//! 游乐场命令：会话的增删改、发一轮对话、出一批图、停、看进行中的。
//!
//! 这一层只做两件事：解出本地网关的地址与口令（`gateway::local_endpoint`），
//! 以及把流式回字变成 `playground://chat` 事件。持久化与历史拼装都在 `nexus-playground`。
//! 图片字节不走 IPC —— 前端凭 id 经 `nexus-image://` 协议取（见 `lib.rs`）。

use crate::commands::events;
use crate::commands::gateway::{local_endpoint, TryFrame};
use crate::state::AppState;
use nexus_core::{AppError, Result};
use nexus_playground::{
    ActiveRun, Asset, Attachment, ImageRequest, Kind, Message, Source, Thread, ThreadDetail,
    ThreadSummary, VideoRequest,
};
use tauri::{AppHandle, Emitter, State};

#[tauri::command(async)]
pub fn playground_threads(
    state: State<'_, AppState>,
    kind: Option<Kind>,
) -> Result<Vec<ThreadSummary>> {
    state.playground.threads(kind)
}

#[tauri::command(async)]
pub fn playground_thread(state: State<'_, AppState>, id: String) -> Result<ThreadDetail> {
    state.playground.thread(&id)
}

#[tauri::command(async)]
pub fn playground_thread_create(
    state: State<'_, AppState>,
    kind: Kind,
    model: String,
) -> Result<Thread> {
    state
        .playground
        .create_thread(kind, Source::Local, &model, None)
}

#[tauri::command(async)]
pub fn playground_thread_rename(
    state: State<'_, AppState>,
    id: String,
    title: String,
) -> Result<Thread> {
    state.playground.rename_thread(&id, &title)
}

/// 换这个会话下一次发送用的模型。
#[tauri::command(async)]
pub fn playground_thread_set_target(
    state: State<'_, AppState>,
    id: String,
    model: String,
) -> Result<Thread> {
    state
        .playground
        .set_target(&id, Source::Local, &model, None)
}

#[tauri::command(async)]
pub fn playground_thread_delete(state: State<'_, AppState>, id: String) -> Result<()> {
    state.playground.delete_thread(&id)
}

#[tauri::command(async)]
pub fn playground_message_delete(state: State<'_, AppState>, id: String) -> Result<()> {
    state.playground.delete_message(&id)
}

/// 这个会话有没有请求正在跑（切走再切回来时接上半截回复）。
#[tauri::command(async)]
pub fn playground_active(
    state: State<'_, AppState>,
    thread_id: String,
) -> Result<Option<ActiveRun>> {
    Ok(state.playground.active(&thread_id))
}

/// 发一轮对话。`prompt` 缺省 = 重新生成末尾那条回复。
///
/// `attachments` 是随这一句发上去的图（base64 进来，落盘后挂在 user 消息上）；缺省为空，
/// 只发字的调用方不必带它。
///
/// 命令本身等流走完才返回，返回的是**已落库**的那条回复；中途的字都在 `playground://chat`
/// 事件里，按 `request_id` 认。
#[tauri::command]
pub async fn playground_chat_send(
    app: AppHandle,
    state: State<'_, AppState>,
    request_id: String,
    thread_id: String,
    prompt: Option<String>,
    attachments: Option<Vec<Attachment>>,
) -> Result<Message> {
    if request_id.trim().is_empty() {
        return Err(AppError::invalid("缺少 request_id。"));
    }
    state.playground.thread(&thread_id)?;
    let ep = local_endpoint(&state)?;
    let id = request_id.clone();
    state
        .playground
        .chat(
            &request_id,
            &thread_id,
            prompt.as_deref(),
            attachments.as_deref().unwrap_or(&[]),
            &ep,
            |event| {
                let _ = app.emit(
                    events::PLAYGROUND_CHAT,
                    TryFrame {
                        id: id.clone(),
                        event,
                    },
                );
            },
        )
        .await
}

/// 停掉一次进行中的请求（对话或出图）。没有这么一次也不算错——用户可能连点了两下。
#[tauri::command(async)]
pub fn playground_stop(state: State<'_, AppState>, request_id: String) -> Result<bool> {
    Ok(state.playground.stop(&request_id))
}

/// 出一批图。同样等做完才返回；图片会话没有流式，进度只有「在跑 / 跑完」。
///
/// 走 `{base}/v1/images/generations`：本地网关那条背后是 Cursor 的 `RunGenerateImage`
/// （固定 1536×1024、账号要有 Developer / Sand 计划的生图权限），权限不够时网关会把原因
/// 写进错误体，这里照样落成一条带 `error` 的回复。
#[tauri::command]
pub async fn playground_image_generate(
    state: State<'_, AppState>,
    request_id: String,
    thread_id: String,
    request: ImageRequest,
) -> Result<Message> {
    if request_id.trim().is_empty() {
        return Err(AppError::invalid("缺少 request_id。"));
    }
    state.playground.thread(&thread_id)?;
    let ep = local_endpoint(&state)?;
    state
        .playground
        .generate_image(&request_id, &thread_id, request, &ep)
        .await
}

/// 出一段视频。上游是异步任务，这里等它做完（最多几分钟）再返回；中途点「停」就不再等。
/// 走 `{base}/v1/videos/*`：只有 Grok 通道有；没有时网关会回 404，落成带 `error` 的回复。
#[tauri::command]
pub async fn playground_video_generate(
    state: State<'_, AppState>,
    request_id: String,
    thread_id: String,
    request: VideoRequest,
) -> Result<Message> {
    if request_id.trim().is_empty() {
        return Err(AppError::invalid("缺少 request_id。"));
    }
    state.playground.thread(&thread_id)?;
    let ep = local_endpoint(&state)?;
    state
        .playground
        .generate_video(&request_id, &thread_id, request, &ep)
        .await
}

/// 资产页：全部生成过的图，新的在前。
#[tauri::command(async)]
pub fn playground_assets(state: State<'_, AppState>) -> Result<Vec<Asset>> {
    state.playground.assets()
}

/// 删一张图（连盘上的文件）。消息本身留着。
#[tauri::command(async)]
pub fn playground_image_delete(state: State<'_, AppState>, id: String) -> Result<()> {
    state.playground.delete_image(&id)
}

/// 在系统文件管理器里显示这张图。
///
/// **必须留在主线程，不要加 `(async)`。** 理由同 `backup_reveal`：`reveal_item_in_dir`
/// 在 macOS 上直接调 AppKit 的 `NSWorkspace`，插件没有自己往主队列派发。
#[tauri::command]
pub fn playground_image_reveal(
    app: AppHandle,
    state: State<'_, AppState>,
    id: String,
) -> Result<()> {
    let (path, _) = state
        .playground
        .image_path(&id)?
        .ok_or_else(|| AppError::invalid("这张图已经不在了。"))?;
    if !path.is_file() {
        return Err(AppError::invalid("这张图的文件已经不在了。")
            .with_hint("它可能被手动删掉了，或这份记录来自别的机器的备份。"));
    }
    use tauri_plugin_opener::OpenerExt;
    app.opener()
        .reveal_item_in_dir(&path)
        .map_err(|e| AppError::internal(format!("打不开文件管理器：{e}")))
}
